### PR TITLE
parameters: small changes to help in the event of a failure

### PR DESCRIPTION
**Pull request by dagar**
Here are a few collected changes to try and improve things in the rare event of a parameter import failure. If param import fails we try to gather data to microsd so we might have some chance of diagnosing. 

``` Console

nsh> param dump
[param] reading from /fs/mtd_params

BSON document size 1287
BSON_DOUBLE: BAT1_A_PER_V = 36.367516
BSON_INT32:  BAT1_SOURCE = -1
BSON_DOUBLE: BAT1_V_DIV = 18.100000
BSON_DOUBLE: BAT2_A_PER_V = 36.367516
BSON_DOUBLE: BAT2_V_DIV = 18.100000
BSON_INT32:  CAL_ACC0_ID = 3670026
BSON_INT32:  CAL_ACC0_PRIO = 50
BSON_DOUBLE: CAL_ACC0_TEMP = 42.757526
BSON_DOUBLE: CAL_ACC0_XOFF = -0.003996
BSON_DOUBLE: CAL_ACC0_YOFF = 0.007201
BSON_DOUBLE: CAL_ACC0_ZOFF = -0.037510
BSON_INT32:  CAL_ACC1_ID = 3932170
BSON_INT32:  CAL_ACC1_PRIO = 50
BSON_DOUBLE: CAL_ACC1_TEMP = 42.478569
BSON_DOUBLE: CAL_ACC1_XOFF = -0.004939
BSON_DOUBLE: CAL_ACC1_YOFF = 0.006842
BSON_DOUBLE: CAL_ACC1_ZOFF = -0.287072
BSON_INT32:  CAL_ACC2_ID = 4259850
BSON_INT32:  CAL_ACC2_PRIO = 50
BSON_DOUBLE: CAL_ACC2_TEMP = 38.000000
BSON_DOUBLE: CAL_ACC2_XOFF = -0.003569
BSON_DOUBLE: CAL_ACC2_YOFF = 0.017583
BSON_DOUBLE: CAL_ACC2_ZOFF = -0.027534
BSON_INT32:  CAL_GYRO0_ID = 3670026
BSON_INT32:  CAL_GYRO0_PRIO = 50
BSON_DOUBLE: CAL_GYRO0_TEMP = 41.151775
BSON_DOUBLE: CAL_GYRO0_XOFF = 0.010774
BSON_DOUBLE: CAL_GYRO0_YOFF = -0.009861
BSON_DOUBLE: CAL_GYRO0_ZOFF = 0.006048
BSON_INT32:  CAL_GYRO1_ID = 3932170
BSON_INT32:  CAL_GYRO1_PRIO = 50
BSON_DOUBLE: CAL_GYRO1_TEMP = 40.618114
BSON_DOUBLE: CAL_GYRO1_XOFF = 0.002919
BSON_DOUBLE: CAL_GYRO1_YOFF = 0.013177
BSON_DOUBLE: CAL_GYRO1_ZOFF = -0.003862
BSON_INT32:  CAL_GYRO2_ID = 4325386
BSON_INT32:  CAL_GYRO2_PRIO = 50
BSON_INT32:  CAL_MAG0_PRIO = 50
BSON_INT32:  CAL_MAG1_PRIO = 50
BSON_INT32:  CBRK_BUZZER = 782090
BSON_DOUBLE: EKF2_MAG_DECL = -6.134731
BSON_INT32:  PWM_AUX_OUT = 1234
BSON_INT32:  PWM_MAIN_MAX1 = 1999
BSON_INT32:  PWM_MAIN_MAX2 = 1999
BSON_INT32:  PWM_MAIN_MAX3 = 1999
BSON_INT32:  PWM_MAIN_MAX4 = 1999
BSON_INT32:  PWM_MAIN_MAX5 = 1999
BSON_INT32:  PWM_MAIN_MAX6 = 1999
BSON_INT32:  PWM_MAIN_MAX7 = 1999
BSON_INT32:  PWM_MAIN_MAX8 = 1999
BSON_INT32:  PWM_MAIN_MIN1 = 1600
BSON_INT32:  PWM_MAIN_MIN2 = 1600
BSON_INT32:  PWM_MAIN_MIN3 = 1600
BSON_INT32:  PWM_MAIN_MIN4 = 1600
BSON_INT32:  PWM_MAIN_MIN5 = 1600
BSON_INT32:  PWM_MAIN_MIN6 = 1600
BSON_INT32:  PWM_MAIN_MIN7 = 1600
BSON_INT32:  PWM_MAIN_MIN8 = 1600
BSON_INT32:  PWM_MAIN_OUT = 1234
BSON_INT32:  SDLOG_DIRS_MAX = 10
BSON_INT32:  SDLOG_MODE = 1
BSON_INT32:  SDLOG_PROFILE = 19
BSON_INT32:  SYS_AUTOSTART = 4001
BSON_EOO
BSON decoded 1287 bytes (double:25, string:0, bin:0, bool:0, int32:38, int64:0)

```

### CODE REVIEW

**Comment by dagar** [comment]
I've added a simple parameter backup mechanism that additionally exports the parameters to `/fs/microsd/parameters_backup.bson` (after FRAM). During init (rcS) if the param import happens to fail we import from the backup file if it exists. This isn't perfect, but it should reduce the pain from an intermittent param import failure in most cases, while also gathering additional data about those failures (to address them directly).

File: src/lib/parameters/parameters.cpp [hunk 1]
```diff
@@ -1404,11 +1451,43 @@ param_import_callback(bson_decoder_t decoder, void *priv, bson_node_t node)
...
+		return 1; // just skip this entry
+	}
+
+	return -1;
+}
+
 static int
 param_import_internal(int fd, bool mark_saved)
 {
 	for (int attempt = 1; attempt < 5; attempt++) {
```
Comment (davids5) [inline src/lib/parameters/parameters.cpp:1489]: Do we still need these attempts?

**Review by davids5 — COMMENTED** [review]
@dagar I would split out the the i32 changes from the functional changes to make this easier to review

File: src/lib/parameters/parameters.cpp (hunk 1 above)
Comment (dagar) [reply src/lib/parameters/parameters.cpp:1489]: > Do we still need these attempts?

Debatable, but right now you're pretty much screwed without parameters so it's worth trying anything.

I'm hoping these changes allow us to capture actual data from these rare failures and then we can re-evaluate.

File: ROMFS/px4fmu_common/init.d/rcS [hunk 2]
```diff
@@ -161,8 +163,31 @@ else
...
+		if [ -f "/fs/microsd/parameters_backup.bson" ]
+		then
+			param import /fs/microsd/parameters_backup.bson
+		fi
 	fi
+
+	if [ $SDCARD_AVAILABLE = yes ]
+	then
+		param select-backup /fs/microsd/parameters_backup.bson
+	fi
```
Comment (julianoes) [inline ROMFS/px4fmu_common/init.d/rcS:189]: Does this always create a backup? How does that work?

File: src/lib/parameters/parameters.cpp [hunk 3]
```diff
@@ -1080,47 +1079,66 @@ param_reset_specific(const char *resets[], int num_resets)
 int
 param_set_default_file(const char *filename)
 {
+	if ((param_backup_file && strcmp(filename, param_backup_file) == 0)) {
+		PX4_ERR("default file can't be the same as the backup file %s", filename);
```
Comment (julianoes) [inline src/lib/parameters/parameters.cpp:1083]: Should you return with -1 here?

File: src/lib/parameters/parameters.cpp [hunk 4]
```diff
@@ -1080,47 +1079,66 @@ param_reset_specific(const char *resets[], int num_resets)
...
 }
 
-int param_save_default()
+int param_set_backup_file(const char *filename)
 {
-	int res = PX4_ERROR;
+	if (param_default_file && strcmp(filename, param_default_file) == 0) {
 
-	const char *filename = param_get_default_file();
+		PX4_ERR("backup file can't be the same as the default file %s", filename);
```
Comment (julianoes) [inline src/lib/parameters/parameters.cpp:1115]: Should you return with -1 here?

File: ROMFS/px4fmu_common/init.d/rcS (hunk 2 above)
Comment (dagar) [reply ROMFS/px4fmu_common/init.d/rcS:189]: At the moment it's just a secondary export file. The primary is typically FRAM (/fs/mtd_params), but could also be another file somewhere (eg /fs/microsd/params).

File: src/lib/parameters/parameters.cpp (hunk 3 above)
Comment (dagar) [reply src/lib/parameters/parameters.cpp:1083]: Thanks, fixed.

File: src/lib/parameters/parameters.cpp (hunk 4 above)
Comment (dagar) [reply src/lib/parameters/parameters.cpp:1115]: Thanks, fixed.

**Review by julianoes — APPROVED** [review]

**Pull request merged**